### ISSUE TITLE
May delete label when using replace  action in relabel_configs

### ISSUE THREAD

**Issue by ddongchen** [prometheus/prometheus#5789]
## Bug Report
**What did you do?**
prometheus config file such as:

scrap_configs:
- job_name: xxx
  relabel_configs:
  - source_labels: [pod]
    action: replace
    target_label: "pod_name"

**What did you expect to see?**
For two metrics, such as:
  metric_normal("test":"aaaa", "pod": "bbbb")
  metric_abnormal("test":"cccc", "pod_name": "dddd")

For "metric_normal", the "pod" should rename to "pod_name", for "metric_abnormal", should keep the "pod_name".

**What did you see instead? Under which circumstances?**
For "metric_normal", the replace action works well, and rename "pod" to "pod_name", but for "metric_abnormal", the “pod_name” label will be dropped. 

I checked the source code,  and found why this is happened. The "replace" action source code is :

	case Replace:
		indexes := cfg.Regex.FindStringSubmatchIndex(val)
		// If there is no match no replacement must take place.
		if indexes == nil {
			break
		}
		target := model.LabelName(cfg.Regex.ExpandString([]byte{}, cfg.TargetLabel, val, indexes))
		if !target.IsValid() {
			lb.Del(cfg.TargetLabel)
			break
		}
		**res := cfg.Regex.ExpandString([]byte{}, cfg.Replacement, val, indexes)**
		if len(res) == 0 {
			lb.Del(cfg.TargetLabel)
			break
		}
		lb.Set(string(target), string(res))
		glog.Infof("after labels: %v", lb.Labels())

For "metric_abnormal", the "val" is an empty string. and "**cfg.Regex.ExpandString([]byte{}, cfg.Replacement, val, indexes)**" return an empty string, then delete the target label. I wonder why should delete the target label?  also shall we just return when "val" is null?

**Comment by brian-brazil** [comment]
Are you sure you're using relabel_configs and not metric_relabel_configs? What version is this?

**Comment by ddongchen** [comment]
I‘m sure using the "relabel_configs".  The version is "2.10", I also tried the source codes from "master" branch, and still have the problems.

**Comment by brian-brazil** [comment]
Can you share exactly what's on the /metrics for these, and this target's row on the Target status page?

**Comment by ddongchen** [comment]
/metrics like:

 curl http://xxx:5678/metric

# HELP container_request_gpu_memory request of gpu memory in MiB
# TYPE container_request_gpu_memory gauge
container_request_gpu_memory{container_name="gpu",namespace="default",node="xxx",pod_name="xxx",req_of_gpu_memory="total"} 2560
container_request_gpu_memory{container_name="gpu",namespace="default",node="xxx",pod_name="xxx",req_of_gpu_memory="total"} 2560

"Target status page", do you mean this?
![image](https://user-images.githubusercontent.com/18438172/61620238-0d0b4580-aca3-11e9-8cb8-b2eea3ca962a.png)

**Comment by ddongchen** [comment]
hi, @brian-brazil . There is an easy way to reproduce the problem. You can collect any metrics, such as  metric_name("label1":"xx", "label2":"xx"), the metric has the label named "label2", then set the config like:

relabel_configs:
- source_labels: [xxx] (xxx must not equal to the value of any label)
   action: replace
   target_label: "label2"

then the "label2" will be dropped.

**Comment by brian-brazil** [comment]
Okay, that'd be a bug then.

**Comment by csmarchbanks** [comment]
Is this a bug? The default regex is `(.*)`, so I would expect matching an empty string/missing label like in PromQL is the desired behaviour.

Either way, adding
```
regex: (.+)
```
to the relabel config makes it explicit that you only want to match when your source_labels have a value.

**Comment by ddongchen** [comment]
hi, @csmarchbanks . Yes, we have added the "regex: (.+)" to match non-empty value, to avoid dropping the label. While this based on we have checked the resource codes and know how it works.

**Comment by ddongchen** [comment]
hi, @brian-brazil,  it is "metric_relabel_configs", not "relabel_configs", so sorry for that!

**Comment by brian-brazil** [comment]
In that case, this is working as expected.

**Comment by jnummelin** [comment]
Is there any tricks to make `metric_relabel_configs` not to drop the existing target label if source label is not found?

**Comment by brian-brazil** [comment]
@jnummelin Please don't ask usage questions on issues as it makes things hard to follow, the prometheus-users mailing list is a better place for them.

**Comment by brian-brazil** [comment]
Going through our bug scrub, this seems to be behaving as expected so closing.
